src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs

```diff
@@ -159,179 +109,283 @@ public IDisposable On(string methodName, Type[] parameterTypes, Func<object[], o
+                Log.FailedToSendInvocation(_logger, invocationMessage.InvocationId, ex);
+                TryRemoveInvocation(invocationMessage.InvocationId, out _);
                 irq.Fail(ex);
-                TryRemoveInvocation(hubMessage.InvocationId, out _);
             }
         }
 
-        public async Task SendAsync(string methodName, object[] args, CancellationToken cancellationToken = default) =>
-            await SendAsyncCore(methodName, args, cancellationToken).ForceAsync();
-
-        private async Task SendAsyncCore(string methodName, object[] args, CancellationToken cancellationToken)
+        private async Task SendHubMessage(HubInvocationMessage hubMessage)
         {
-            if (!_startCalled)
-            {
-                throw new InvalidOperationException($"The '{nameof(SendAsync)}' method cannot be called before the connection has been started.");
-            }
+            AssertConnectionValid();
 
-            var invocationMessage = new InvocationMessage(null, target: methodName,
-                argumentBindingException: null, arguments: args);
+            var payload = _protocol.WriteToArray(hubMessage);
 
-            ThrowIfConnectionTerminated(invocationMessage.InvocationId);
+            Log.SendingInvocation(_logger, hubMessage.InvocationId);
+            await WriteAsync(payload);
+            Log.InvocationSent(_logger, hubMessage.InvocationId);
+        }
+
+        private async Task SendAsyncCore(string methodName, object[] args)
+        {
+            CheckDisposed();
 
+            await _startLock.WaitAsync();
             try
             {
+                CheckDisposed();
+                if (_connection == null)
+                {
+                    throw new InvalidOperationException($"The '{nameof(SendAsync)}' method cannot be called if the connection is not active");
+                }
+
                 Log.PreparingNonBlockingInvocation(_logger, methodName, args.Length);
 
-                var payload = _protocol.WriteToArray(invocationMessage);
-                Log.SendInvocation(_logger, invocationMessage.InvocationId);
+                var invocationMessage = new InvocationMessage(null, target: methodName,
+                    argumentBindingException: null, arguments: args);
 
-                await _connection.SendAsync(payload, cancellationToken);
-                Log.SendInvocationCompleted(_logger, invocationMessage.InvocationId);
+                await SendHubMessage(invocationMessage);
             }
-            catch (Exception ex)
+            finally
             {
-                Log.SendInvocationFailed(_logger, invocationMessage.InvocationId, ex);
-                throw;
+                _startLock.Release();
             }
         }
 
-        private async Task OnDataReceivedAsync(byte[] data)
+        private async Task<(bool close, Exception exception)> ProcessMessagesAsync(ReadOnlySequence<byte> buffer)
```

Comment:
tuples...

Reply:
internal method and `out` doesn't work because async.